enzyme/test/Enzyme/ProbProg/condition-static.ll

```diff
@@ -0,0 +1,204 @@
+; CHECK-NEXT: entry:
+; CHECK-NEXT:   %trace1 = call i8* @__enzyme_newtrace()
+; CHECK-NEXT:   %0 = call i1 @__enzyme_has_choice(i8* %trace, i8* getelementptr inbounds ([2 x i8], [2 x i8]* @.str.1, i64 0, i64 0))
+; CHECK-NEXT:   br i1 %0, label %condition.call.with.trace, label %condition.call.without.trace
+
+; CHECK: condition.call.with.trace:                        ; preds = %entry
+; CHECK-NEXT:   %1 = alloca double
+; CHECK-NEXT:   %2 = bitcast double* %1 to i8*
+; CHECK-NEXT:   %3 = call i64 @__enzyme_get_choice(i8* %trace, i8* getelementptr inbounds ([2 x i8], [2 x i8]* @.str.1, i64 0, i64 0), i8* %2, i64 8)
+; CHECK-NEXT:   %4 = load double, double* %1
+; CHECK-NEXT:   br label %entry.cntd
+
+; CHECK: condition.call.without.trace:                     ; preds = %entry
+; CHECK-NEXT:   %sample.call = call double @normal(double 0.000000e+00, double 1.000000e+00)
+; CHECK-NEXT:   br label %entry.cntd
+
+; CHECK: entry.cntd:                                       ; preds = %condition.call.without.trace, %condition.call.with.trace
+; CHECK-NEXT:   %5 = phi double [ %4, %condition.call.with.trace ], [ %sample.call, %condition.call.without.trace ]
+; CHECK-NEXT:   %likelihood.call = call double @normal_logpdf(double 0.000000e+00, double 1.000000e+00, double %5)
+; CHECK-NEXT:   %6 = alloca double
+; CHECK-NEXT:   store double %5, double* %6
+; CHECK-NEXT:   %7 = bitcast double* %6 to i8**
+; CHECK-NEXT:   %8 = load i8*, i8** %7
+; CHECK-NEXT:   call void @__enzyme_insert_choice(i8* %trace1, i8* getelementptr inbounds ([2 x i8], [2 x i8]* @.str.1, i64 0, i64 0), double %likelihood.call, i8* %8, i64 8)
+; CHECK-NEXT:   %9 = call i1 @__enzyme_has_choice(i8* %trace, i8* getelementptr inbounds ([2 x i8], [2 x i8]* @.str.2, i64 0, i64 0))
+; CHECK-NEXT:   br i1 %9, label %condition.call1.with.trace, label %condition.call1.without.trace
+
+; CHECK: condition.call1.with.trace:                       ; preds = %entry.cntd
+; CHECK-NEXT:   %10 = alloca double
+; CHECK-NEXT:   %11 = bitcast double* %10 to i8*
+; CHECK-NEXT:   %12 = call i64 @__enzyme_get_choice(i8* %trace, i8* getelementptr inbounds ([2 x i8], [2 x i8]* @.str.2, i64 0, i64 0), i8* %11, i64 8)
+; CHECK-NEXT:   %13 = load double, double* %10
+; CHECK-NEXT:   br label %entry.cntd.cntd
+
+; CHECK: condition.call1.without.trace:                    ; preds = %entry.cntd
+; CHECK-NEXT:   %sample.call1 = call double @normal(double 0.000000e+00, double 1.000000e+00)
+; CHECK-NEXT:   br label %entry.cntd.cntd
+
+; CHECK: entry.cntd.cntd:                                  ; preds = %condition.call1.without.trace, %condition.call1.with.trace
+; CHECK-NEXT:   %14 = phi double [ %13, %condition.call1.with.trace ], [ %sample.call1, %condition.call1.without.trace ]
+; CHECK-NEXT:   %likelihood.call1 = call double @normal_logpdf(double 0.000000e+00, double 1.000000e+00, double %14)
+; CHECK-NEXT:   %15 = alloca double
```

Comment:
This alloca should be at the entry block [or rather the equivalent of inversionAllocs]. This is because if the allocation is made inside of a loop of size N, it will continue to push the stack pointer each iteration and potentially overflow the stack, rather than the intended behavior.

Reply:
fair